File: rgr/events/EventEmitter.py
from collections.abc import Callable
from typing import Any

type Callback = Callable[[list[Any]], None]
type Listeners = dict[str, list[Callback]]

class EventEmitter:

  def __init__(self):
    self.listeners: Listeners = {}

  def on(self, name: str, listener: Callback):
    listeners = self.listeners.get(name, [])
    listeners.append(listener)
    self.listeners[name] = listeners
    return self

  def emit(self, name: str, *args):
    listeners = self.listeners.get(name, [])
    for listener in listeners: listener(args)
    return self

  def removeAllListeners(self, name):
    if name in self.listeners: self.listeners.pop(name)
    return self

  def removeListener(self, name: str, listener: Callback):
    listeners = self.listeners.get(name, [])
    listeners.remove(listener)
    return self

  def once(self, name: str, listener: Callback):
    def wrapped(*args):
      listener(args)
      self.listeners.get(name, []).remove(wrapped)
    self.on(name, wrapped)
    return self

File: rgr/events/EventEmitter.java
package events;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class EventEmitter<T> {
  private final Map<String, List<Consumer<T>>> listeners = new HashMap<>();

  public EventEmitter<T> on(final String name, final Consumer<T> listener) {
    final var exists = listeners.containsKey(name);
    if (exists) listeners.get(name).add(listener);
    else listeners.put(name, new ArrayList<>(List.of(listener)));
    return this;
  }

  public EventEmitter<T> emit(final String name, final T data) {
    if (!listeners.containsKey(name)) return this;
    for (final var listener: listeners.get(name)) {
      listener.accept(data);
    }
    return this;
  }

  public EventEmitter<T> off(final String name) {
    if (listeners.containsKey(name)) listeners.remove(name);
    return this;
  }

  public EventEmitter<T> off(final String name, final Consumer<T> listener) {
    if (listeners.containsKey(name)) {
      final var list = listeners.get(name);
      if (list.contains(listener)) list.remove(listener);
    }
    return this;
  }
}
